File: UTPC/2024_10_23_Contest_WiCS/houdini.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <map>
#include <set>
#include <stack>
#include <functional>
#include <cmath>
#include <numeric>
#include <iomanip>

using namespace std;
typedef long long ll;
typedef long double ld;
#define sz(x) (int)(x).size()
#define all(x) begin(x), end(x)

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	int n;
	cin >> n;
	vector<int> p(n);
	for(int i = 0; i < n; i++) {
		cin >> p[i];
		p[i] %= 2;
	}
	vector<vector<int>> con(n);
	for(int i = 0; i < n-1; i++) {
		int a, b;
		cin >> a >> b;
		a--; b--;
		con[a].push_back(b);
		con[b].push_back(a);
	}
	
	int ans = 1;
	
	function<int(int, int)> dfs = [&](int curr, int prev) {
		int tot = p[curr];
		for(int next : con[curr]) {
			if(next != prev)
				tot += dfs(next, curr);
		}
		tot %= 2;
		if(tot==0 && curr != 0)
			ans++;
		return tot;
	};
	
	if(dfs(0, -1) == 1)
		ans = -1;
	cout << ans << "\n";
	
}

File: UTPC/2024_10_23_Contest_WiCS/houdini.java
import java.util.*;
import java.io.*;

public class houdini {
	
	static int ans = 1;
	
	public static int dfs(int curr, int prev, ArrayList<Integer>[] con, int[] v) {
		int tot = v[curr];
		for(int next : con[curr])
			if(next != prev)
				tot+=dfs(next, curr, con, v);
		tot%=2;
		if(tot%2==0 && curr != 0)
			ans++;
		return tot;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(in.readLine());
		StringTokenizer st = new StringTokenizer(in.readLine());
		int[] v = new int[n];
		ArrayList<Integer>[] con = new ArrayList[n];
		for(int i = 0; i < n; i++) {
			v[i] = Integer.parseInt(st.nextToken())%2;
			con[i] = new ArrayList<>();
		}
		
		for(int i = 0; i < n-1; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken())-1;
			int b = Integer.parseInt(st.nextToken())-1;
			con[a].add(b);
			con[b].add(a);
		}
		if(dfs(0, -1, con, v) == 1)
			ans = -1;
		System.out.println(ans);
	}
}